### PR TITLE
 Remove logic for guessing slug from an unregistered domain

### LINKED ISSUE #1991: Domain resolution too clever
One of the tactics that we use for domain magic can result in some unexpected problems. If the project does not have a domain associated with it yet, our middleware will attempt a DNS lookup. This block will attempt to get a slug from a CNAME:
https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/core/middleware.py#L67-L76

We aren't doing anything to check that the domain is our though. In this case, the domain `virtualenv.pypa.io` resolves as such:

```
% dig +short virtualenv.pypa.io
c.global-ssl.fastly.net.
199.27.79.175
```

The above block will split the CNAME returned, assume the slug is the `c`, and apply the domain to this project if it exists.

Instead, we should probably ensure the CNAME domain is something we control, and this is indeed a slug, not a hostname.

### PR DESCRIPTION
**Pull request by stsewd**
I messed up #4730 trying to solve the conflicts. See the another PR for the discussion (sorry!)

Close #1991

### CODE REVIEW

**Comment by humitos** [comment]
We can remove the `dnspython` dependency now, https://github.com/rtfd/readthedocs.org/blob/49134eef84be5cff2cb259c6aac436db41d99fc5/requirements/pip.txt#L44

**Comment by ericholscher** [comment]
I think this is ready to go, but I think we need to do a bit more work to reduce the possible downsides. We have some manual migration steps, but any project that is missed will randomly have their Domain stop working. This is not a great experience, so I think we should:

* Create a specific 404/error page for missing Domain entries, which gives users more details on the error and links to the docs on how to fix it
* Log all invalid domains somewhere, perhaps Redis, with a counter, and then log an error if we hit some reasonable number (10, 20?) to Sentry, so that we can catch "real" domains rather than typos.

File: readthedocs/core/templates/core/dns-404.html [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+<h3>{% blocktrans %}The host "{{ host }}" is unknown to Read the Docs{% endblocktrans %}</h3>
+
+<p>
+  {% with docsurl='https://docs.readthedocs.io/en/stable/custom_domains.html' %}
+  {% blocktrans trimmed %}
+    If you control this domain and believe this is in error,
+    please review our <a href="{{ docsurl }}">custom domain documentation</a>.
+    In the past, we allowed custom domains to point to us without configuring the domain in the Read the Docs dashboard
+    and we attempted to intelligently guess the correct project.
+    This was never explicitly supported and documented. It just happened to work.
```
Comment (ericholscher) [inline readthedocs/core/templates/core/dns-404.html]: Believe it used to be documented.

Comment (davidfischer) [reply readthedocs/core/templates/core/dns-404.html]: I can update that then

**Comment by davidfischer** [comment]
> Log all invalid domains somewhere, perhaps Redis, with a counter, and then log an error if we hit some reasonable number (10, 20?) to Sentry, so that we can catch "real" domains rather than typos.

We have logs over the last 3 months. I added the domain where the following were true:

- DNS pointed to readthedocs
- I could reasonably verify that the domain was controlled by the project
- A page requiring the DNS happened 20+ times in a month (60x in the last 90 days)

**Comment by stsewd** [comment]
Going to fix the tests

File: readthedocs/core/templates/core/dns-404.html [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+{% block content %}
+<h1>{% trans '404 - Invalid Host' %}</h1>
+<h3>{% blocktrans %}The host "{{ host }}" is unknown to Read the Docs{% endblocktrans %}</h3>
+
+<p>
+  {% with docsurl='https://docs.readthedocs.io/en/stable/custom_domains.html' %}
+  {% blocktrans trimmed %}
+    If you control this domain and believe this is in error,
```
Comment (stsewd) [inline readthedocs/core/templates/core/dns-404.html]: Shouldn't be `is an error`?

```suggestion
    If you control this domain and believe this is an error,
```

Comment (davidfischer) [reply readthedocs/core/templates/core/dns-404.html]: Either works but I think yours is better.

**Pull request merged**